source/authentication/apple.txt

```diff
@@ -59,379 +59,607 @@ following:
+            .. code-block:: ruby
 
+               # Source: https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple
+               require 'jwt'
 
-   .. step:: Create the Client Secret JWT
+               # Update these values with your app's information
+               team_id = '<Apple Team ID>'
+               client_id = '<Bundle ID>'
+               key_id = '<Key ID>'
+               key_file = '<Key File Path>'
 
-      You can now create the client secret JWT for the Apple ID
-      authentication provider. Make sure that you have the following
-      information:
+               # Define the JWT's headers and claims
+               headers = {
+               # The token must be signed with your key
+               'kid' => key_id
+               }
+               claims = {
+               # The token is issued by your Apple team
+               'iss' => team_id,
+               # The token applies to Apple ID authentication
+               'aud' => 'https://appleid.apple.com',
+               # The token is scoped to your application
+               'sub' => client_id,
+               # The token is valid immediately
+               'iat' => Time.now.to_i,
+               # The token expires in 6 months (maximum allowed)
+               'exp' => Time.now.to_i + 86400*180,
+               }
 
-      - The :guilabel:`Services ID` that you created.
+               # Read in the key and generate the JWT
+               ecdsa_key = OpenSSL::PKey::EC.new IO.read key_file
+               token = JWT.encode claims, ecdsa_key, 'ES256', headers
 
-      - The :guilabel:`Key ID` of the key that you created and the ``.p8``
-        file that contains the key.
+               # Print the JWT to stdout
+               puts token
 
-      - Your Apple Team ID. You can find this in the top right of the Apple
-        Developer Portal.
+            Update the values of ``team_id``, ``client_id``, ``key_id``, and
+            ``key_file`` to match your application's information and then save
+            the file. When you're ready to generate the JWT, run the script in
+            your shell:
 
-        .. figure:: /images/apple-team-id.png
-           :alt: An Apple team ID in the Apple Developer Portal
+            .. code-block:: shell
 
-      Once you've confirmed that you have all the required information, you can use
-      a script to generate the JWT. You may define your own script or use the
-      script in this step.
+               ruby generate_client_secret.rb >> client_secret.txt
 
-      To generate the JWT, we'll use the :rubygems:`jwt <gems/jwt>` gem. To install
-      it, run the following:
+            .. important:: Save the JWT
 
-      .. code-block:: shell
+               The ``generate_client_secret.rb`` script saves the JWT string to a
+               file called ``client_secret.txt``. You will need the JWT to
+               configure the Apple ID provider in App Services.
```

Comment:
Not quite true - the script outputs to stdout but when we call it we append stdout to `client_secret.txt`

Reply:
Hah, this is literally just a copy-paste of what was already there. But sure, updating!

Reply:
Yes well a younger me was technically incorrect, which is the worst kind!